### PR TITLE
//seq.trim_{prefix,suffix}

### PR DESCRIPTION
**Pull request by marcelocantos**
Fixes # .

Changes proposed in this pull request:
- 
- 

Checklist:
- [x] Added related tests
- [x] Made corresponding changes to the documentation

### CODE REVIEW

File: docs/std-seq.md [hunk 1]
```diff
@@ -99,3 +99,29 @@ Usage:
 | example | equals |
 |:-|:-|
 | `//seq.repeat(2, "hots")` | `"hotshots"` |
+
+## `//seq.trim_prefix(prefix <: array, subject <: array) <: array` <br/> `//seq.trim_prefix(prefix <: string, subject <: string) <: string`
```
Comment (nofun97) [inline docs/std-seq.md:103]: what about `Bytes`? the same with `trim_suffix`

File: syntax/std_seq.go [hunk 2]
```diff
@@ -226,15 +226,47 @@ func stdSeqSplit(delimiter, subject rel.Value) rel.Value {
 	panic(fmt.Errorf("split: unsupported args: %s, %s", delimiter, subject))
 }
 
+func stdSeqTrimPrefix(prefix, subject rel.Value) rel.Value {
```
Comment (nofun97) [inline syntax/std_seq.go:229]: would this offset the sequence?

File: syntax/std_seq_array_helper.go [hunk 3]
```diff
@@ -145,15 +144,35 @@ func arrayHasSuffix(suffix rel.Value, subject rel.Array) rel.Value {
...
-func convert2Array(val rel.Value) rel.Array {
-	switch val := val.(type) {
-	case rel.Array:
-		return val
-	case rel.GenericSet:
-		valArray, _ := rel.AsArray(val)
-		return valArray
+func arrayTrimPrefix(prefix rel.Value, subject rel.Array) rel.Value {
+	if prefix.IsTrue() && subject.IsTrue() && arrayHasPrefix(prefix, subject).IsTrue() {
+		if prefixLen := prefix.(rel.Array).Count(); prefixLen <= subject.Count() {
```
Comment (nofun97) [inline syntax/std_seq_array_helper.go:149]: this check doesn't seem necessary, if `arrayHasPrefix` is true, then `prefixLen <= subject.Count()` is always true

File: syntax/std_seq_array_helper.go [hunk 4]
```diff
@@ -145,15 +144,35 @@ func arrayHasSuffix(suffix rel.Value, subject rel.Array) rel.Value {
...
+		}
 	}
+	return subject
+}
 
+func arrayTrimSuffix(suffix rel.Value, subject rel.Array) rel.Value {
+	if suffix.IsTrue() && subject.IsTrue() && arrayHasSuffix(suffix, subject).IsTrue() {
+		suffixLen := suffix.(rel.Array).Count()
+		subjectLen := subject.Count()
+		if suffixLen <= subjectLen {
```
Comment (nofun97) [inline syntax/std_seq_array_helper.go:165]: ditto

File: syntax/std_seq_prefix_test.go [hunk 5]
```diff
@@ -63,3 +63,65 @@ func TestBytesPrefix(t *testing.T) {
...
+	AssertCodesEvalToSameValue(t, `"ABCDE"`, `//seq.trim_prefix("BCD","ABCDE")`)
+	AssertCodesEvalToSameValue(t, `"ABCDE"`, `//seq.trim_prefix("CD","ABCDE")`)
+	AssertCodesEvalToSameValue(t, `""`, `//seq.trim_prefix("CD","")`)
+	AssertCodesEvalToSameValue(t, `"ABCD"`, `//seq.trim_prefix("","ABCD")`)
+
+	AssertCodesEvalToSameValue(t, `""`, `//seq.trim_prefix("","")`)
+	AssertCodesEvalToSameValue(t, `""`, `//seq.trim_prefix("A","")`)
+	AssertCodesEvalToSameValue(t, `"A"`, `//seq.trim_prefix("","A")`)
+
+	AssertCodeErrors(t, `//seq.trim_prefix(1,"ABC")`, "")
```
Comment (nofun97) [inline syntax/std_seq_prefix_test.go:82]: what's the error message?

File: syntax/std_seq_prefix_test.go [hunk 6]
```diff
@@ -63,3 +63,65 @@ func TestBytesPrefix(t *testing.T) {
...
+
+	AssertCodesEvalToSameValue(t, `[3]`, `//seq.trim_prefix([1, 2],[1, 2, 3])`)
+	AssertCodesEvalToSameValue(t, `[[3]]`, `//seq.trim_prefix([[1, 2]],[[1, 2], [3]])`)
+
+	AssertCodesEvalToSameValue(t, `[]`, `//seq.trim_prefix([], [])`)
+	AssertCodesEvalToSameValue(t, `[]`, `//seq.trim_prefix(['A','B','C','D','E','F'],[])`)
+	AssertCodesEvalToSameValue(t, `['A','B','C','D','E']`, `//seq.trim_prefix([],['A','B','C','D','E'])`)
+
+	AssertCodeErrors(t, `//seq.trim_prefix(1,[1,2,3])`, "")
+	AssertCodeErrors(t, `//seq.trim_prefix('A',['A','B','C'])`, "")
```
Comment (nofun97) [inline syntax/std_seq_prefix_test.go:111]: ditto

File: syntax/std_seq_prefix_test.go [hunk 7]
```diff
@@ -63,3 +63,65 @@ func TestBytesPrefix(t *testing.T) {
...
+	AssertCodesEvalToSameValue(t, `        ['D','E']`, `//seq.trim_prefix(['A','B','C']    ,['A','B','C','D','E'])`)
+	AssertCodesEvalToSameValue(t, `            ['E']`, `//seq.trim_prefix(['A','B','C','D'],['A','B','C','D','E'])`)
+	AssertCodesEvalToSameValue(t, `['A','B','C','D','E']`, `//seq.trim_prefix(['B'],['A','B','C','D','E'])`)
+	AssertCodesEvalToSameValue(t, `['A','B','C','D','E']`, `//seq.trim_prefix(['B','C'],['A','B','C','D','E'])`)
+	AssertCodesEvalToSameValue(t,
+		`['A','B','C','D','E']`,
+		`//seq.trim_prefix(['A','B','C','D','E','F'],['A','B','C','D','E'])`)
+	AssertCodesEvalToSameValue(t,
+		`['A','B','C','D','E']`,
+		`//seq.trim_prefix(['A','B','C','D','E','F'],['A','B','C','D','E'])`)
```
Comment (nofun97) [inline syntax/std_seq_prefix_test.go:101]: these two are the same tests

File: syntax/std_seq_prefix_test.go [hunk 8]
```diff
@@ -63,3 +63,65 @@ func TestBytesPrefix(t *testing.T) {
...
+
+	AssertCodesEvalToSameValue(t, `[]`, `//seq.trim_prefix([], [])`)
+	AssertCodesEvalToSameValue(t, `[]`, `//seq.trim_prefix(['A','B','C','D','E','F'],[])`)
+	AssertCodesEvalToSameValue(t, `['A','B','C','D','E']`, `//seq.trim_prefix([],['A','B','C','D','E'])`)
+
+	AssertCodeErrors(t, `//seq.trim_prefix(1,[1,2,3])`, "")
+	AssertCodeErrors(t, `//seq.trim_prefix('A',['A','B','C'])`, "")
+}
+
+func TestBytesTrimPrefix(t *testing.T) {
```
Comment (nofun97) [inline syntax/std_seq_prefix_test.go:114]: add `AssertCodeErrors`

File: syntax/std_seq_suffix_test.go [hunk 9]
```diff
@@ -57,3 +57,61 @@ func TestBytesSuffix(t *testing.T) {
...
+	AssertCodesEvalToSameValue(t, `"ABCD"`, `//seq.trim_suffix("E","ABCDE")`)
+	AssertCodesEvalToSameValue(t, `"ABC"`, `//seq.trim_suffix("DE","ABCDE")`)
+	AssertCodesEvalToSameValue(t, `"ABCDE"`, `//seq.trim_suffix("CD", "ABCDE")`)
+	AssertCodesEvalToSameValue(t, `"ABCDE"`, `//seq.trim_suffix("D","ABCDE")`)
+
+	AssertCodesEvalToSameValue(t, `""`, `//seq.trim_suffix("D","")`)
+	AssertCodesEvalToSameValue(t, `""`, `//seq.trim_suffix("","")`)
+	AssertCodesEvalToSameValue(t, `"ABCDE"`, `//seq.trim_suffix("","ABCDE")`)
+
+	AssertCodeErrors(t, `//seq.trim_suffix(1,"ABC")`, "")
```
Comment (nofun97) [inline syntax/std_seq_suffix_test.go:74]: ditto

File: syntax/std_seq_prefix_test.go (hunk 5 above)
Comment (nofun97) [reply syntax/std_seq_prefix_test.go:82]: actually it's a bit strange, the trim function doesn't seem to return error

File: syntax/std_seq_suffix_test.go [hunk 10]
```diff
@@ -57,3 +57,61 @@ func TestBytesSuffix(t *testing.T) {
...
+	AssertCodesEvalToSameValue(t, `[[1 ,2]]`, `//seq.trim_suffix([3, 4],[[1 ,2], 3, 4])`)
+	AssertCodesEvalToSameValue(t, `[]      `, `//seq.trim_suffix([3, 4],[3, 4])`)
+	AssertCodesEvalToSameValue(t, `[[1, 2]]`, `//seq.trim_suffix([[3, 4]],[[1 ,2], [3, 4]])`)
+
+	AssertCodesEvalToSameValue(t, `['A','B','C','D','E']`, `//seq.trim_suffix([],['A','B','C','D','E'])`)
+	AssertCodesEvalToSameValue(t, `[]`, `//seq.trim_suffix([], [])`)
+	AssertCodesEvalToSameValue(t, `[]`, `//seq.trim_suffix(['D','E'],[])`)
+
+	AssertCodeErrors(t, `//seq.trim_suffix(1,[1,2])`, "")
+	AssertCodeErrors(t, `//seq.trim_suffix('A',['A','B'])`, "")
```
Comment (nofun97) [inline syntax/std_seq_suffix_test.go:101]: ditto

File: syntax/std_seq_suffix_test.go [hunk 11]
```diff
@@ -57,3 +57,61 @@ func TestBytesSuffix(t *testing.T) {
...
+}
+
+func TestBytesTrimSuffix(t *testing.T) {
```
Comment (nofun97) [inline syntax/std_seq_suffix_test.go:104]: add `AssertCodeErrors`

File: docs/std-seq.md (hunk 1 above)
Comment (marcelocantos) [reply docs/std-seq.md:103]: Good point. I copy-pasted another section that is also missing bytes.

File: syntax/std_seq_prefix_test.go (hunk 5 above)
Comment (nofun97) [reply syntax/std_seq_prefix_test.go:81]: ah there's `AssertExprPanics`, but it's weird that `AssertCodeErrors` doesn't fail

File: syntax/std_seq_prefix_test.go (hunk 6 above)
Comment (nofun97) [reply syntax/std_seq_prefix_test.go:107]: sorry, meant it's just like this https://github.com/arr-ai/arrai/pull/358#discussion_r433047905

File: syntax/std_seq_suffix_test.go (hunk 9 above)
Comment (nofun97) [reply syntax/std_seq_suffix_test.go:74]: same with this https://github.com/arr-ai/arrai/pull/358#discussion_r433047905

File: syntax/std_seq_suffix_test.go (hunk 10 above)
Comment (nofun97) [reply syntax/std_seq_suffix_test.go:101]: same with this https://github.com/arr-ai/arrai/pull/358#discussion_r433047905

File: syntax/std_seq_prefix_test.go (hunk 5 above)
Comment (marcelocantos) [reply syntax/std_seq_prefix_test.go:81]: `wrapPanics` converts it to an error.

Comment (marcelocantos) [reply syntax/std_seq_prefix_test.go:81]: I'm not a big fan of validating error messages, tbh. The important thing is that it fails.

**Pull request merged**